### PR TITLE
review(front): refacto url-manager

### PR DESCRIPTION
**Pull request by AMoreaux**
Replace https://github.com/twentyhq/twenty/pull/8855

### CODE REVIEW

File: packages/twenty-front/src/utils/recoil-effects.ts [hunk 1]
```diff
@@ -20,6 +21,20 @@ export const localStorageEffect =
...
+const customCookieAttributeZodSchema = z.object({
+  cookieAttributes: z.object({
+    expires: z.union([z.number(), z.instanceof(Date)]).optional(),
+    path: z.string().optional(),
+    domain: z.string().optional(),
+    secure: z.boolean().optional(),
+  }),
+});
+
+export const isCustomCookiesAttributesValue = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/utils/recoil-effects.ts:33]: ty!

File: packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { isDefined } from '~/utils/isDefined';
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { useRecoilValue } from 'recoil';
+import { isMultiWorkspaceEnabledState } from '@/client-config/states/isMultiWorkspaceEnabledState';
+import { useDefaultDomain } from '@/domain-manager/hooks/useDefaultDomain';
+
+export const useWorkspaceSubdomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts:7]: useReadWorkspaceSubdomainFromCurrentLocation

File: packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  const domainConfiguration = useRecoilValue(domainConfigurationState);
+  const isMultiWorkspaceEnabled = useRecoilValue(isMultiWorkspaceEnabledState);
+  const { defaultDomain } = useDefaultDomain();
+
+  const isWorkspaceSubdomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts:12]: isOnAWorkspaceSubdomain

File: packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      !isDefined(domainConfiguration.frontDomain) ||
+      !isDefined(domainConfiguration.defaultSubdomain)
+    ) {
+      throw new Error('frontDomain and defaultSubdomain are required');
+    }
+
+    return window.location.hostname !== defaultDomain;
+  };
+
+  const workspaceSubdomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts:25]: subdomain

File: packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts (hunk 2 above)
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useWorkspaceSubdomain.ts:7]: const location = useLocation()


const isOnAWorkspaceSubdomain = location.hostname !== defaultDomain
const workspaceSubdomain = location.replace(...)

return { isOnAWorkspaceSubdomain, workspaceSubdomain }

File: packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+export const useDefaultDomain = () => {
+  const domainConfiguration = useRecoilValue(domainConfigurationState);
+  const isMultiWorkspaceEnabled = useRecoilValue(isMultiWorkspaceEnabledState);
+
+  const defaultDomain = isMultiWorkspaceEnabled
+    ? `${domainConfiguration.defaultSubdomain}.${domainConfiguration.frontDomain}`
+    : domainConfiguration.frontDomain;
+
+  const isDefaultDomain = isMultiWorkspaceEnabled
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts:13]: use location = useLocation()

File: packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { isMultiWorkspaceEnabledState } from '@/client-config/states/isMultiWorkspaceEnabledState';
+import { useRecoilValue } from 'recoil';
+
+export const useDefaultDomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts:5]: useIsCurrentLocationOnDefaultDomain
+ useDefaultDomain

File: packages/twenty-front/src/modules/domain-manager/hooks/useBuildWorkspaceUrl.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { isDefined } from '~/utils/isDefined';
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { useRecoilValue } from 'recoil';
+
+export const useBuildWorkspaceUrl = () => {
+  const domainConfiguration = useRecoilValue(domainConfigurationState);
+
+  return (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useBuildWorkspaceUrl.ts:8]: const myFunction = () => {}

return { nameFunction: myFunction }

File: packages/twenty-front/src/modules/domain-manager/hooks/useBuildWorkspaceUrl.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    subdomain?: string,
+    onPage?: string,
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useBuildWorkspaceUrl.ts:10]: pathname

File: packages/twenty-front/src/modules/domain-manager/hooks/useDomainBackToDefaultSubdomain.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { useDefaultDomain } from '@/domain-manager/hooks/useDefaultDomain';
+
+export const useDomainBackToDefaultSubdomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useDomainBackToDefaultSubdomain.ts:3]: useRedirectToDefaultDomain

Comment (charlesBochet) [reply packages/twenty-front/src/modules/domain-manager/hooks/useDomainBackToDefaultSubdomain.ts:3]: same return comment

File: packages/twenty-front/src/modules/domain-manager/hooks/useDomainBackToWorkspace.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { isMultiWorkspaceEnabledState } from '@/client-config/states/isMultiWorkspaceEnabledState';
+import { useRecoilValue } from 'recoil';
+import { useBuildWorkspaceUrl } from '@/domain-manager/hooks/useBuildWorkspaceUrl';
+
+export const useDomainBackToWorkspace = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useDomainBackToWorkspace.ts:5]: useRedirectToWorkspaceDomain

same return comment

File: packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import { useRecoilCallback, useRecoilValue } from 'recoil';
+import { lastAuthenticateWorkspaceDomainState } from '@/domain-manager/states/lastAuthenticateWorkspaceDomainState';
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { isDefined } from '~/utils/isDefined';
+
+export const useLastAuthenticateWorkspaceDomain = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts:6]: useLastAuthenticatedWorkspaceDomain

File: packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+                  domain: `.${domainConfiguration.frontDomain}`,
+                },
+              }
+            : null,
+        );
+      },
+    [domainConfiguration],
+  );
+
+  const lastAuthenticateWorkspaceDomain = useRecoilValue(
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts:28]: don't read lastAuthenticateWorkspaceDomain. we consider states to be part of the api of our modules

File: packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+import { lastAuthenticateWorkspaceDomainState } from '@/domain-manager/states/lastAuthenticateWorkspaceDomainState';
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { isDefined } from '~/utils/isDefined';
+
+export const useLastAuthenticateWorkspaceDomain = () => {
+  const domainConfiguration = useRecoilValue(domainConfigurationState);
+
+  const setLastAuthenticateWorkspaceDomain = useRecoilCallback(
+    ({ set }) =>
+      (params: { id: string; subdomain: string } | null) => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts:11]: workspaceId

File: packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import { useRecoilCallback, useRecoilValue } from 'recoil';
+import { lastAuthenticateWorkspaceDomainState } from '@/domain-manager/states/lastAuthenticateWorkspaceDomainState';
+import { domainConfigurationState } from '@/domain-manager/states/domainConfigurationState';
+import { isDefined } from '~/utils/isDefined';
+
+export const useLastAuthenticateWorkspaceDomain = () => {
+  const domainConfiguration = useRecoilValue(domainConfigurationState);
+
+  const setLastAuthenticateWorkspaceDomain = useRecoilCallback(
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts:9]: from the caller side

If you want:

Case 1:
const setLastAuthenticatedWorkspaceDomain = useLastAuthenticateWorkspaceDomain()
...
setLastAuthenticatedWorkspaceDomain({ workspaceId, subdomain })
==> you'll need to use a recoilcallback

Comment (charlesBochet) [reply packages/twenty-front/src/modules/domain-manager/hooks/useLastAuthenticateWorkspaceDomain.ts:9]: const setLastAuthenticatedWorkspaceDomain = useSetRecoilState()

File: packages/twenty-front/src/modules/domain-manager/states/lastAuthenticateWorkspaceDomainState.ts [hunk 15]
```diff
@@ -2,16 +2,16 @@ import { cookieStorageEffect } from '~/utils/recoil-effects';
 import { Workspace } from '~/generated/graphql';
 import { createState } from 'twenty-ui';
 
-export const lastAuthenticateWorkspaceState = createState<
+export const lastAuthenticateWorkspaceDomainState = createState<
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/states/lastAuthenticateWorkspaceDomainState.ts:5]: authenticated

File: packages/twenty-front/src/modules/domain-manager/states/lastAuthenticateWorkspaceDomainState.ts [hunk 16]
```diff
@@ -2,16 +2,16 @@ import { cookieStorageEffect } from '~/utils/recoil-effects';
...
       cookieAttributes?: Cookies.CookieAttributes;
     })
   | null
 >({
-  key: 'lastAuthenticateWorkspaceState',
+  key: 'lastAuthenticateWorkspaceDomain',
   defaultValue: null,
   effects: [
-    cookieStorageEffect('lastAuthenticateWorkspace', {
+    cookieStorageEffect('lastWorkspaceDomainState', {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/domain-manager/states/lastAuthenticateWorkspaceDomainState.ts:14]: lastAuthenticateWorkspaceDomainState (state?)

File: packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts (hunk 5 above)
Comment (AMoreaux) [reply packages/twenty-front/src/modules/domain-manager/hooks/useDefaultDomain.ts:13]: useLocation doesn't expose the hostname.

**Review by charlesBochet — APPROVED** [review]
LGTM!

**Pull request merged**